### PR TITLE
Java: add some neutral models discovered with heuristics

### PR DESCRIPTION
**Pull request by jcogs33**
#### Description
This PR adds some neutral models discovered with heuristics. 
The `apache` model was discovered with a ssrf heuristic. The rest of the models were discovered with a path-injection heuristic.

(cc @atorralba @tiferet)

#### Consideration
- I put a provenance of `"manual"` on these since they were manually reviewed after initially being generated by heuristic queries. Let me know if I should use a different provenance for heuristic-generated models (related to [this discussion](https://github.com/github/codeql/pull/12228#discussion_r1109981958)). 
- I didn't add a change-note because I don't think neutral models are really a "user-visible" change. Let me know if I'm wrong about that and if I should add a change note for these models.
 - I haven't added test cases for neutral models before, but let me know if I should add some "tests" that just have a `// no flow` comment or something similar.

### CODE REVIEW

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 1]
```diff
@@ -34,3 +34,22 @@ extensions:
...
+      - ["java.nio.file", "Files" "isExecutable", "", "manual"]
+      - ["java.nio.file", "Files" "isHidden", "", "manual"]
+      - ["java.nio.file", "Files" "isReadable", "", "manual"]
+      - ["java.nio.file", "Files" "isRegularFile", "", "manual"]
+      - ["java.nio.file", "Files" "isSameFile", "", "manual"]
+      - ["java.nio.file", "Files" "isSymbolicLink", "", "manual"]
+      - ["java.nio.file", "Files" "isWritable", "", "manual"]
+      - ["java.nio.file", "Files" "notExists", "", "manual"]
+      - ["java.nio.file", "Files" "setLastModifiedTime", "", "manual"]
+      - ["java.nio.file", "Files" "size", "", "manual"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:55]:
```suggestion
      - ["java.nio.file", "Files", "exists", "", "manual"]
      - ["java.nio.file", "Files", "getLastModifiedTime", "", "manual"]
      - ["java.nio.file", "Files", "getOwner", "", "manual"]
      - ["java.nio.file", "Files", "getPosixFilePermissions", "", "manual"]
      - ["java.nio.file", "Files", "isDirectory", "", "manual"]
      - ["java.nio.file", "Files", "isExecutable", "", "manual"]
      - ["java.nio.file", "Files", "isHidden", "", "manual"]
      - ["java.nio.file", "Files", "isReadable", "", "manual"]
      - ["java.nio.file", "Files", "isRegularFile", "", "manual"]
      - ["java.nio.file", "Files", "isSameFile", "", "manual"]
      - ["java.nio.file", "Files", "isSymbolicLink", "", "manual"]
      - ["java.nio.file", "Files", "isWritable", "", "manual"]
      - ["java.nio.file", "Files", "notExists", "", "manual"]
      - ["java.nio.file", "Files", "setLastModifiedTime", "", "manual"]
      - ["java.nio.file", "Files", "size", "", "manual"]
```

**Review by atorralba — COMMENTED** [review]
Looks good! I think there was a small formatting mistake I added some suggestions for.

> I put a provenance of "manual" on these since they were manually reviewed after initially being generated by heuristic queries. Let me know if I should use a different provenance for heuristic-generated models (related to https://github.com/github/codeql/pull/12228#discussion_r1109981958).

I think `manual` is the right provenance according to the discussion.

> I didn't add a change-note because I don't think neutral models are really a "user-visible" change. Let me know if I'm wrong about that and if I should add a change note for these models.

This doesn't need a change note indeed.

> I haven't added test cases for neutral models before, but let me know if I should add some "tests" that just have a `// no flow` comment or something similar.

For neutral summaries, it probably doesn't make sense to add tests. But for neutral sinks, which are manually added and from my POV have a stronger and more intentional sense of "this is safe/non-exploitable", I think it could make sense to have tests — so that we get notified if that assumption is no loner valid and someone adds such a sink in the future.

File: java/ql/lib/ext/java.nio.file.model.yml (hunk 1 above)
Comment (jcogs33) [reply java/ql/lib/ext/java.nio.file.model.yml]: Thanks for catching these typos! Updated in https://github.com/github/codeql/pull/12249/commits/0b99aa0bd3538944a517b82ca38bd7ee72e95bd5.

**Comment by jcogs33** [comment]
> For neutral summaries, it probably doesn't make sense to add tests. But for neutral sinks, which are manually added and from my POV have a stronger and more intentional sense of "this is safe/non-exploitable", I think it could make sense to have tests — so that we get notified if that assumption is no loner valid and someone adds such a sink in the future.

Sounds good, I'll add tests for these, but I'll wait until the discussion from elsewhere about exactly how to represent neutral sinks is finalized. 
Converting this PR back to a draft until then.

**Comment by jcogs33** [comment]
Re-opening this PR for review now that https://github.com/github/codeql/pull/12931 is completed.
As a reminder, this PR is just adding a few neutral sink models that were discovered when reviewing heuristic results, it is not intended to be complete.

Changes since this was last reviewed:
 - added "sink" kinds
 - updated the provenance to "hq-manual"
 - added tests as was recommended above

Regarding the tests, let me know if you think there is a better way to test these. The test case I've added just uses [`sinkNode`](https://github.com/github/codeql/blob/5bc844c4c643370d030edd1222e447c1d3954500/java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll#L469) to check if a MaD sink exists and will fail if a MaD sink is added for one of the neutrals, but it currently won't fail if a sink is added in regular CodeQL. Let me know if there is a good way to include a check for sinks written in regular CodeQL.
Also, would it make sense to have a more general check/test to make sure that sinks and neutral sinks never overlap? Maybe something like `getInvalidNeutralModel` added to [`invalidModelRow`](https://github.com/github/codeql/blob/5bc844c4c643370d030edd1222e447c1d3954500/java/ql/lib/semmle/code/java/dataflow/ExternalFlow.qll#L337)?

**Comment by atorralba** [comment]
Models themselves LGTM.

> Let me know if there is a good way to include a check for sinks written in regular CodeQL.

I guess that would only work for queries with extensible sinks (i.e. abstract classes), and you'd need to explicitly use them in tests, which doesn't seem ideal. I think we're fine testing only models-as-data sinks for the moment.

On a different note, you're currently testing that certain callables don't contain sinks, but not that we actually have neutral models for them. Do you think it makes sense to also test that with something like the following?

```ql
pragma[nomagic]
predicate isNeutral() { this = any(FlowSummaryImpl::Public::NeutralCallable nsc).asCallable() }
```

Let me know if you think it's overkill.

**Comment by jcogs33** [comment]
> I guess that would only work for queries with extensible sinks (i.e. abstract classes), and you'd need to explicitly use them in tests, which doesn't seem ideal. I think we're fine testing only models-as-data sinks for the moment.

👍 

> On a different note, you're currently testing that certain callables don't contain sinks, but not that we actually have neutral models for them. Do you think it makes sense to also test that with something like the following?
> 
> ```ql
> pragma[nomagic]
> predicate isNeutral() { this = any(FlowSummaryImpl::Public::NeutralCallable nsc).asCallable() }
> ```
> 
> Let me know if you think it's overkill.

It seems somewhat overkill to me, but I don't have a strong opinion about it 🙂, so I added something to test for this in https://github.com/github/codeql/pull/12249/commits/82f208ca7a113ebd04276a1011ec6336307b6882.  Note that `FlowSummaryImpl::Public::NeutralCallable` is specific to [neutral **summaries**](https://github.com/github/codeql/blob/4497aa5eb11f5e7bbdee0dbd0989f2b49ee4a8da/java/ql/lib/semmle/code/java/dataflow/internal/FlowSummaryImpl.qll#L331), so I wasn't able to use it to test for neutral sinks. What I did instead feels a bit hacky to me, so let me know if you think it isn't worth adding this or if you think there's a better way to write this test case. (I suppose I could create a `NeutralSinkCallable` class similar to `FlowSummaryImpl::Public::NeutralCallable`, but I was trying to keep it simple by just using existing classes/predicates).

**Pull request merged**